<?xml version="1.0"?>
<launch>
  <node name="lidar_imu_calibration" pkg="lidar_imu_calibration" type="lidar_imu_calibration" output="screen" clear_params="true" required="true">

    <!-- bag topics -->
    <param name="topic_imu"    type="string" value="/vectornav/IMU"/>
    <param name="topic_lidar"  type="string" value="/os_cloud_node/points"/>

    <!-- bag parameters -->
    <param name="path_bag" type="string" value="$(find lidar_imu_calibration)/bag/2021-04-16-10-32-05_far.bag"/>
    <param name="bag_start"   type="double" value="0.0"/>
    <param name="bag_durr"    type="int"    value="-1"/>

    <!-- Estimator Parameters -->
    <param name="use_fej" type="bool" value="false"/>
    <param name="use_imuavg" type="bool" value="true"/>
    <param name="use_rk4int" type="bool" value="true"/>
    <param name="calib_lidar_timeoffset" type="bool" value="false"/>
    <param name="calib_extrinsics"   type="bool"   value="true"/>
    <param name="max_clones" type="int" value="2"/>
    <param name="updater_do_chi2_check" type="bool" value="false"/>

    <!-- Filter Init Parameters -->
    <param name="init_window_time" type="double" value="0.5"/> <!-- in second -->
    <param name="init_imu_thresh" type="double" value="0.25"/> <!-- jerk threshold -->

    <!-- Noise Parameters -->
    <param name="accelerometer_noise_density"  type="double"   value="0.0015825749257056074"/>
    <param name="accelerometer_random_walk"    type="double"   value="4.1557212532459854e-05"/>
    <param name="gyroscope_noise_density"      type="double"   value="8.97991030686202e-05"/>
    <param name="gyroscope_random_walk"        type="double"   value="7.253101866424753e-07"/>

    <!-- Updater Parameters -->
    <!-- Chi2 Parameters -->
    <param name="updater_chi2_multiplier" type="int" value="1"/>
    <!-- Update Noise (sigma) Params -->
    <param name="updater_rotation_noise" type="double" value="0.1"/>
    <param name="updater_translation_noise" type="double" value="0.1"/>

    <!-- State Initial Noise (sigma) Params -->
    <param name="state_init_x_noise" type="double" value="0.5"/>
    <param name="state_init_y_noise" type="double" value="0.5"/>
    <param name="state_init_z_noise" type="double" value="0.5"/>
    <param name="state_init_rx_noise" type="double" value="0.1"/>
    <param name="state_init_ry_noise" type="double" value="0.1"/>
    <param name="state_init_rz_noise" type="double" value="0.1"/>
    <param name="state_init_timeoffset_noise" type="double" value="0.01"/>

    <!-- Gravity -->
    <rosparam param="gravity">[0.0, 0.0, 9.81]</rosparam>

    <!-- initialization of lidar imu time offset -->
    <param name="calib_lidarimu_dt" type="double" value="0.0"/>

    <!-- params determining map size -->
    <param name="limit_map_size" type="bool" value="false"/>
    <param name="no_of_scans_for_map" type="int" value="400"/>

    <!-- params for downsampling clouds before mapping -->
    <param name="downsample_for_mapping" type="bool" value="false"/>

    <!-- initial calibration file -->
    <param name="init_lidar_inertial_calibration_result_filename" type="string" value="$(find lidar_imu_calibration)/data/I_T_L_init.txt"/>

    <!-- output filenames -->
    <param name="lidar_odometry_trajectory_filename" type="string" value="$(find lidar_imu_calibration)/data/lidar_odometry_trajectory"/>
    <param name="inertial_trajectory_filename" type="string" value="$(find lidar_imu_calibration)/data/inertial_trajectory"/>
    <param name="inertial_bias_filename" type="string" value="$(find lidar_imu_calibration)/data/inertial_bias"/>
    <param name="inertial_velocity_filename" type="string" value="$(find lidar_imu_calibration)/data/inertial_velocity"/>
    <param name="lidar_inertial_calib_extrinsic_filename" type="string" value="$(find lidar_imu_calibration)/data/lidar_inertial_calib_extrinsic"/>
    <param name="lidar_inertial_calib_dt_filename" type="string" value="$(find lidar_imu_calibration)/data/lidar_inertial_calib_dt"/>

    <!-- NDT Resolution -->
    <param name="ndt_resolution" type="double" value="0.25"/>

    <!-- Undistortion flag-->
    <param name="do_undistortion" type="bool" value="true"/>

    <!-- Write an output -->
    <param name="lidar_inertial_calibration_result_filename" type="string" value="$(find lidar_imu_calibration)/data/I_T_L_final.txt"/>
    <param name="gen_map_data" type="bool" value="true"/>
    <param name="map_csv_file_name" type="string" value="$(find lidar_imu_calibration)/data/map.csv"/>
  </node>

</launch>
